netket/sampler/rules/hamiltonian.py

```diff
@@ -114,3 +116,70 @@ def _choose(vp, sections, rand_vec, out, w):
         out[i] = vp[n_rand]
         w[i] = math.log(s - low_range)
         low_range = s
+
+
+@struct.dataclass
+class HamiltonianRuleJax(HamiltonianRule):
+    """
+    Rule proposing moves according to the terms in an operator.
+
+    In this case, the transition matrix is taken to be:
+
+    .. math::
+
+       T( \\mathbf{s} \\rightarrow \\mathbf{s}^\\prime) = \\frac{1}{\\mathcal{N}(\\mathbf{s})}\\theta(|H_{\\mathbf{s},\\mathbf{s}^\\prime}|),
+
+    This rule only works with operators which are written in jax.
+    """
+
+    def transition(self, _0, _1, _2, _3, key, x):
+
+        xp, mels = self.operator.get_conn_padded(x)
+
+        n_conn = self.operator.n_conn(x)
+
+        # def _n_conn(mels):
+        #     nonzeros = jnp.abs(mels) > 0
+        #     return nonzeros.sum(axis=-1)
+        # n_conn = _n_conn(mels)
+
+        rand_i = jax.random.randint(key, shape=(x.shape[0],), minval=0, maxval=n_conn)
```

Comment:
here it would be useful to have  some kind of `JaxOperator.get_conn_padded_and_n_conn` as in some of the operators to come we compute the n_conn inside of get_conn_padded anyway.

(here n_conn tells us how many nonzeros there are at the beginning of the mels of each sample, the rest is assumed to be zero padding, which would not have been present in the old get_conn_flattened)